reader-shp/src/main/java/com/graphhopper/reader/shp/GraphhopperShp.java

```diff
@@ -0,0 +1,26 @@
+package com.graphhopper.reader.shp;
+
+import java.util.HashSet;
+
+import com.graphhopper.GraphHopper;
+import com.graphhopper.reader.DataReader;
+import com.graphhopper.reader.shp.OSMShapeFileReader.EdgeAddedListener;
+import com.graphhopper.storage.GraphHopperStorage;
+
+public class GraphhopperShp extends GraphHopper {
+	private final HashSet<EdgeAddedListener> edgeAddedListeners = new HashSet<>();
+
+	@Override
+	protected DataReader createReader(GraphHopperStorage ghStorage) {
+		OSMShapeFileReader reader = new OSMShapeFileReader(ghStorage);
+		for (EdgeAddedListener l : edgeAddedListeners) {
+			reader.addListener(l);
+		}
+		return initDataReader(reader);
+	}
+
+	public void addListener(EdgeAddedListener l) {
```

Comment:
Did I understood the purpose of the listeners: adding custom attributes to the graph? Why are multiple listeners necessary?